### PR TITLE
created and tested bindings for hyperreduction

### PR DESCRIPTION
**Pull request by pravija12345**
### `S_OPT`

The input argument `std::vector<int> *init_samples` cannot have a default value `nullptr`. See the pybind11 [documentation](https://pybind11.readthedocs.io/en/stable/advanced/functions.html#allow-prohibiting-none-arguments).

Although, `init_samples` needs not be a pointer in `S_OPT` implementation. Therefore, we switched it into `std::vector<int> &`, and set the default value to be `std::vector<int>(0)`. This does not impact the internal execution.

The arguments `f_sampled_row` and `f_sampled_rows_per_proc` are STL-type output variables, which cannot change its value if it is passed as an input argument. Therefore it is removed from the input argument list, and the python function return these two variables as tuple:
```
import pylibROM.hyperreduction as hyperreduction
f_sampled_row, f_sampled_rows_per_proc = hyperreduction.S_OPT(u, num_basis_vectors, f_basis_sampled_inv, myid, num_procs, num_samples)
```
Note that `Matrix& f_basis_sampled_inv` is both input and output, and it is not a STL-type. This allows us passing it as an input argument which can modify its value within the function.

### CODE REVIEW

**Comment by dreamer2368** [comment]
Based on the pybind11 [documentation](https://pybind11.readthedocs.io/en/stable/advanced/cast/overview.html), copy is the default action for STL-type arguments. This means that input/output variables of type `std::vector<T>` will essentially always involve a deep copy, unless we set a custom casting action.

@chldkdtn , do you have an idea about the typical size of `std::vector<int>` variables passed into libROM's `hyperreduction` routines, such as `DEIM`, `GNAT`, `S_OPT`, `QDEIM`? If they are not so large, then we may not need to do anything particular.

**Comment by dreamer2368** [comment]
This is for a future reference. If we do not want to have the 'out-of-box' copy action for `std::vector`, we have to construct `py::list` or `py::array_t` and return them, not directly returning `std::vector`. See this [post](https://stackoverflow.com/questions/54876346/pybind11-and-stdvector-how-to-free-data-using-capsules).

**Comment by chldkdtn** [comment]
> Based on the pybind11 [documentation](https://pybind11.readthedocs.io/en/stable/advanced/cast/overview.html), copy is the default action for STL-type arguments. This means that input/output variables of type `std::vector<T>` will essentially always involve a deep copy, unless we set a custom casting action.
> 
> @chldkdtn , do you have an idea about the typical size of `std::vector<int>` variables passed into libROM's `hyperreduction` routines, such as `DEIM`, `GNAT`, `S_OPT`, `QDEIM`? If they are not so large, then we may not need to do anything particular.

Ideally, we do not want it to be big because the whole purpose of the hyper-reduction is to be able to represent a big size of nonlinear vector function with a small number of samples, say 50 or 100 (or perhaps in worst case 1000 at maximum). By the way, this deep copy is only for the output index only, right? which will be part of offline training cost. This should be okay for now, but let's keep it in mind so that in the future, in the context of "on-the-fly," we might need to avoid deep copy to make the efficiency optimized.

File: bindings/pylibROM/python_utils/cpp_utils.hpp [hunk 1]
```diff
@@ -51,4 +51,22 @@ T* getVectorPointer(py::array_t<T> &u_in)
     return static_cast<T*>(buf_info.ptr);
 }
 
+inline int* getVectorPointer_int(py::array_t<int> &u_in)
```
Comment (dreamer2368) [inline bindings/pylibROM/python_utils/cpp_utils.hpp]: I don't think you need to define this for `int*`. The `getVectorPointer` above should be applicable as a template.

File: bindings/pylibROM/hyperreduction/pyQDEIM.cpp [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+namespace py = pybind11;
+using namespace CAROM;
+
+void init_QDEIM(pybind11::module_ &m) {
+   m.def("QDEIM", [](const Matrix* f_basis,int num_f_basis_vectors_used,
+                  Matrix& f_basis_sampled_inv,const int myid,const int num_procs,const int num_samples_req) {
+      std::vector<int> f_sampled_row(num_samples_req);
+      std::vector<int> f_sampled_rows_per_proc(num_procs);
+      QDEIM(f_basis, num_f_basis_vectors_used,f_sampled_row, f_sampled_rows_per_proc,f_basis_sampled_inv, myid, num_procs,num_samples_req);
+      return std::make_tuple(f_sampled_row, f_sampled_rows_per_proc);
```
Comment (dreamer2368) [inline bindings/pylibROM/hyperreduction/pyQDEIM.cpp]: `std::make_tuple` copies the objects, unless it is specified with move. We should use
```
std::make_tuple(std::move(f_sampled_row), std::move(f_sampled_rows_per_proc));
```

File: bindings/pylibROM/hyperreduction/__init__.py [hunk 3]
```diff
@@ -0,0 +1,6 @@
+from .pyDEIM import DEIM
+from .pyGNAT import GNAT
+from .pyQDEIM import QDEIM
+from .pyS_OPT import S_OPT
+from .pySTSampling import STSampling
+from .pyUtilities import Utilities
```
Comment (dreamer2368) [inline bindings/pylibROM/hyperreduction/__init__.py]: This should have only one line:
```
from _pylibROM.hyperreduction import *
```

**Review by dreamer2368 — APPROVED** [review]

**Review by bumpwy — APPROVED** [review]

**Comment by pravija12345** [comment]
Created bindings for the following classes in hyper reduction module:

- DEIM 
- GNAT 
- QDEIM
- S_OPT
- STSampling
- Utilities

In these classes a key modification has been made to the init_samples parameter. The input argument std::vector<int> *init_samples cannot have a default value nullptr, so, we switched it into std::vector<int> &, and set the default value to be std::vector<int>(0) (Note: These changes doesn't impact the internal execution).

Further more another key modification is made to the parameters f_sampled_row, f_sampled_rows_per_proc, and t_samples which were originally passed as input arguments. As these arguments f_sampled_row, f_sampled_rows_per_proc, and t_samples are STL-type output variables, which cannot change its value if it is passed as an input argument. Therefore it is removed from the input argument list, and the python function returns these variables as tuple if necessary.

**Pull request merged**